ostd/src/cpu/cpu_local.rs

```diff
@@ -0,0 +1,223 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! CPU local storage.
+//!
+//! This module provides a mechanism to define CPU-local objects.
+//!
+//! This is acheived by placing the CPU-local objects in a special section
+//! `.cpu_local`. The bootstrap processor (BSP) uses the objects linked in this
+//! section, and these objects are copied to dynamically allocated local
+//! storage of each application processors (AP) during the initialization
+//! process.
+//!
+//! Such a mechanism exploits the fact that constant values of non-[`Copy`]
+//! types can be bitwise copied. For example, a [`Option<T>`] object, though
+//! being not [`Copy`], have a constant constructor [`Option::None`] that
+//! produces a value that can be bitwise copied to create a new instance.
+//! [`alloc::sync::Arc`] however, don't have such a constructor, and thus cannot
+//! be directly used as a CPU-local object. Wrapping it in a type that has a
+//! constant constructor, like [`Option<T>`], can make it CPU-local.
+
+use core::ops::Deref;
+
+use crate::{
+    cpu::{get_cpu_local_base, set_cpu_local_base},
+    trap::{disable_local, DisabledLocalIrqGuard},
+};
+
+/// Defines a CPU-local variable.
+///
+/// # Example
+///
+/// ```rust
+/// use crate::cpu_local;
+/// use core::cell::RefCell;
+///
+/// cpu_local! {
+///     static FOO: RefCell<u32> = RefCell::new(1);
+///
+///     #[allow(unused)]
+///     pub static BAR: RefCell<f32> = RefCell::new(1.0);
+/// }
+///
+/// println!("FOO VAL: {:?}", *FOO.borrow());
+/// ```
+#[macro_export]
+macro_rules! cpu_local {
+    ($( $(#[$attr:meta])* $vis:vis static $name:ident: $t:ty = $init:expr; )*) => {
+        $(
+            #[link_section = ".cpu_local"]
+            $(#[$attr])* $vis static $name: $crate::CpuLocal<$t> = $crate::CpuLocal::__new($init);
+        )*
+    };
+}
+
+extern "C" {
+    fn __cpu_local_start();
+    fn __cpu_local_end();
```

Comment:
Move this block of code to the bottom of this file. They are implementation details.